More general integration of selfing generations

cheuerde opened the pull request:
Some general updates:

- pedigree() function will now run editPed itself if necessary
- pedigree S4 class has 3 more slots: generation, selfing_generation, expanded
- function added: `getGeneration` which returns the generation number. It calls the new C function `get_generation`
- function added: `expandPedigreeSelfing` - based on the selfing_generation, will add appropriate generations to pedigree.
  -  this is based on `getASelfing` logic. 
  - Calls C function `expand_pedgiree_selfing`
  - while all expanded entries will have the suffix "-Fx" (if sepChar = "-F"), but the original individuals wont! 
- `getA` with `labs`  argument now uses `Matrix::crossprod(relfactor(ped))[, labs]` to compute A subset
- `getASubset` now same as getA
- `getASelfing` function still there but now uses the generic pedigree generation

cheuerde committed:
init

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
du

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
ud

cheuerde committed:
sort pedigree if necessary

cheuerde committed:
ud

cheuerde committed:
ud

rramadeu committed:
Update pedigree.R

rramadeu committed:
Update DESCRIPTION

rramadeu committed:
Update pedigree.R

rramadeu committed:
Update NAMESPACE

rramadeu committed:
Merge pull request #1 from rramadeu/patch-1

Update NAMESPACE

rramadeu committed:
Update pedigree.R

rramadeu committed:
Update NAMESPACE

rramadeu committed:
Update DESCRIPTION

rramadeu committed:
Update pedigree.R

cheuerde committed:
Merge pull request #1 from rramadeu/master

minors on namespace

gregorgorjanc commented on R/pedigree.R:
```diff
@@ -421,67 +550,23 @@ getAInv <- function(ped) {
 #' stopifnot(Matrix::isSymmetric(A))
 getA <- function(ped, labs = NULL) {
+
+    L = relfactor(ped)
```
@cheuerde we don't want this for a large pedigree - the `getASubset()` uses the indirect Colleau method that gives a subset of A without ever forming the full L matrix,

cheuerde committed:
fix(pedigree): use original character sire/dam in Level 2 error handler

The tryCatch error handler in pedigree() was calling editPed() with
sire/dam that had already been converted to integer indices by the
try body. editPed then treated these integers as character IDs,
found no matches in the label set, and added them all as phantom
founder entries. This broke ancestor connections and caused zero
coancestry in the NRM.

Fix: save original character sire/dam/label before integer conversion
and use those in the error handler.

cheuerde committed:
docs(pedigree): fix sire/dam param types to match actual usage

sire and dam are character label vectors, not integer positions.
The old docs were inherited from pedigreemm and did not reflect
how the code actually works.

cheuerde committed:
perf(editPed): use C get_generation instead of recursive R getGenAncestors

Replace the O(n^2) pure R recursive loop in editPed with the existing
C get_generation function. On a 434k record pedigree this reduces
editPed from hours to seconds.

cheuerde committed:
perf: dramatic speedup for selfing-expanded pedigrees

Three optimizations that reduce total processing time from hours to seconds:

1. inbreeding() C code: Add selfing shortcut — when sire==dam,
   F = (1 + F_parent)/2 in O(1) instead of tracing all ancestors.
   This skips ~80% of individuals in selfing-expanded pedigrees.
   (1M pedigree: 8+ hours → 0.03 seconds)

2. expandPedigreeSelfing() R wrapper: Replace ped2DF() + factor()
   with direct slot access + match() for O(n) label-to-index mapping.
   Avoids creating expensive factor columns on million-row data.
   (1M pedigree: minutes → 3.5 seconds)

3. pedigree() constructor + editPed(): Replace all factor() calls
   with match() for O(n) label mapping instead of O(n*m).

Full pipeline benchmark (250k input → 1.05M expanded):
  pedigree():                0.28s (was hours)
  expandPedigreeSelfing():   3.6s  (was minutes)
  inbreeding():              0.03s (was 8+ hours)
  TOTAL:                     3.9s
